File: ChainBridge/api/schemas/shadow_pilot.py
"""Pydantic schemas for Shadow Pilot APIs."""

from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict


class ShadowPilotSummaryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    run_id: str
    prospect_name: str
    period_months: int
    total_gmv_usd: float
    financeable_gmv_usd: float
    financed_gmv_usd: float
    protocol_revenue_usd: float
    working_capital_saved_usd: float
    losses_avoided_usd: float
    salvage_revenue_usd: float
    average_days_pulled_forward: float
    shipments_evaluated: int
    shipments_financeable: int
    input_filename: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[datetime] = None


class ShadowPilotShipmentResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    shipment_id: str
    corridor: Optional[str] = None
    mode: Optional[str] = None
    customer_segment: Optional[str] = None
    cargo_value_usd: float
    event_truth_score: float
    eligible_for_finance: bool
    financed_amount_usd: float
    days_pulled_forward: int
    wc_saved_usd: float
    protocol_revenue_usd: float
    avoided_loss_usd: float
    salvage_revenue_usd: float
    exception_flag: bool
    loss_flag: bool


class ShadowPilotShipmentsPageResponse(BaseModel):
    items: List[ShadowPilotShipmentResultResponse]
    next_cursor: Optional[str] = None


class ShadowPilotIngestRequest(BaseModel):
    run_id: Optional[str] = None
    prospect_name: str
    period_months: int
    input_filename: Optional[str] = None
    notes: Optional[str] = None

File: ChainBridge/chainboard-ui/src/types/shadowPilot.ts
/**
 * Shadow Pilot Types
 *
 * TypeScript definitions for Shadow Pilot commercial impact analysis.
 * These mirror Cody's backend API contract for pilot run results.
 */

export interface ShadowPilotSummary {
  run_id: string;
  prospect_name: string;
  period_months: number;
  total_gmv_usd: number;
  financeable_gmv_usd: number;
  financed_gmv_usd: number;
  protocolRevenueUsd: number;
  working_capital_saved_usd: number;
  losses_avoided_usd: number;
  salvageRevenueUsd: number;
  average_daysPulledForward: number;
  shipments_evaluated: number;
  shipments_financeable: number;
  createdAt: string;
}

export interface ShadowPilotRun {
  run_id: string;
  prospect_name: string;
  corridor?: string;
  period_start: string;
  period_end: string;
  status: 'running' | 'completed' | 'failed';
  createdAt: string;
  completed_at?: string;
}

export interface ShadowPilotShipmentResult {
  shipmentId: string;
  original_gmv_usd: number;
  financeable_amount_usd: number;
  financedAmountUsd: number;
  protocol_fee_usd: number;
  working_capital_saved_usd: number;
  daysPulledForward: number;
  riskScore: number;
  financing_applied: boolean;
  reason?: string;
}

export interface ShadowPilotFilters {
  prospect?: string;
  corridor?: string;
  period_months?: number;
  min_gmv?: number;
  status?: 'running' | 'completed' | 'failed';
}

export interface ShadowPilotExportData {
  summary: ShadowPilotSummary;
  shipments?: ShadowPilotShipmentResult[];
  export_timestamp: string;
  exported_by: string;
}

// Derived metrics for UI display
export interface ShadowPilotMetrics {
  total_commercial_impact: number;
  roi_percentage: number;
  financing_penetration: number;
  avg_capital_velocity: number;
  loss_prevention_rate: number;
}
